### ISSUE TITLE
StringIndexOutOfBoundsException when running plugin with watchedResource for top directory

### ISSUE THREAD

**Issue by andersaaberg** [grails/grails-core#10709]
### Steps to Reproduce

1. Run the command: `grails-create-plugin test-plugin`
2. Add a high level directory as watch directory at TestPluginGrailsPlugin.groovy:     
e.g. `def watchedResources = "file:/"` or `def watchedResources = "file:./**/*.properties"`
3. Run `./gradlew bootRun` from the test-plugin directory

### Expected Behaviour
Grails application running at http://localhost:8080 in environment: development

### Actual Behaviour
```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -28
        at java.lang.String.substring(String.java:1875)
        at java_lang_String$substring$2.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at grails.boot.GrailsApp.enableDevelopmentModeWatch(GrailsApp.groovy:194)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:93)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:388)
        at grails.boot.GrailsApp.run(GrailsApp.groovy:375)
        at grails.boot.GrailsApp$run.call(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
        at watchplugin.Application.main(Application.groovy:10)
```
### Environment Information

- **Operating System**: Ubuntu 16.04.2 LTS x64
- **Grails Version:** 3.3.0.BUILD-SNAPSHOT (the issue is also present at version 3.2.9)
- **JDK Version:** OpenJDK 1.7.0_80 x64

### Example Application
https://github.com/andersaaberg/watchPlugin

### FIX
I have identified a fix and will submit a pull request ASAP. The issue is that grails.boot.GrailsApp.groovy has the following code to find the watched file/directory based on the baseDirPath:
```
def locationLength = baseDirPath.length() + 1
[...]
if(wp.file) {
    def resolvedPath = new File(watchBase, wp.file.canonicalPath.substring(locationLength))
    directoryWatcher.addWatchFile(resolvedPath)
}
else if(wp.directory && wp.extension) {
    def resolvedPath = new File(watchBase, wp.directory.canonicalPath.substring(locationLength))
    directoryWatcher.addWatchDirectory(resolvedPath, wp.extension)
}
```

As far as I can see, this can just be replaced by (the pattern with the minus operator is also used elsewhere in the same file):
```
if(wp.file) {
    String relativePath = wp.file.canonicalPath - baseDirPath
    def resolvedPath = new File(watchBase, relativePath)
    directoryWatcher.addWatchFile(resolvedPath)
}
else if(wp.directory && wp.extension) {
    String relativePath = wp.directory.canonicalPath - baseDirPath
    def resolvedPath = new File(watchBase, relativePath)
    directoryWatcher.addWatchDirectory(resolvedPath, wp.extension)
}
```

**Comment by andersaaberg** [comment]
I have made the following pull request: https://github.com/grails/grails-core/pull/10710

I am happy to make some automatic tests for this fix - should that be in grails-core/src/test/groovy/grails/boot ?

**Comment by jeffbrown** [comment]
This looks like it is probably a bug.  Thanks for the feedback.  I am curious... what is the use case for wanting to watch all resources below the top level directory?

**Comment by andersaaberg** [comment]
@jeffbrown: I am upgrading an inhouse legacy plugin (from Grails 2 to Grails 3) that reads key-value pairs from .properties files and it must detect when the .properties files are changed/added and update the key-values accordingly. I cannot make assumptions about the folder names, because the plugin has always been looking at top level and many applications are currently using it.

@graemerocher : I have made another pull request for grails:3.2.x to match the milestone you set: https://github.com/grails/grails-core/pull/10718

**Comment by andersaaberg** [comment]
I have added a unittest: https://github.com/andersaaberg/grails-core/blob/f958466263cbd422c42059e0d2f350f066f43b0b/grails-core/src/test/groovy/grails/boot/DevelopmentModeWatchSpec.groovy

Is there anything else I can do to help getting the pull request (https://github.com/grails/grails-core/pull/10718) ready?

(It seems like the buildserver is having trouble with JAVA_HOME.)

**Comment by andersaaberg** [comment]
The pull request (#10718) does not fail anymore as I have become a member af grails-core on GitHub and fixed the java issue. Is there any more that I need to do to get that pull request included in 3.2.12?

**Comment by andersaaberg** [comment]
Solved in https://github.com/grails/grails-core/commit/b9307ccb3973d3460e5a0ea80e6d03e37cd23f23
